iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/selector/impl/NewSizeTieredCompactionSelector.java

```diff
@@ -242,7 +306,8 @@ private void endCurrentTaskSelection() {
           long currentFileSize = resource.getTsFileSize();
           if (totalFileSize + currentFileSize > singleFileSizeThreshold
               || totalFileNum + 1 > totalFileNumUpperBound
-              || !isFileLevelSatisfied(resource.getTsFileID().getInnerCompactionCount())) {
+              || !isFileLevelSatisfied(resource.getTsFileID().getInnerCompactionCount())
+              || estimateCompactionTaskMemoryDuringSelection) {
```

Comment:
Better to add some comment about why `estimateCompactionTaskMemoryDuringSelection` would affect the judgement (and the one below).
It took me a while to think it through.